From 28753711f02c4f6811e8fb783440effffeac53b5 Mon Sep 17 00:00:00 2001
From: Stefan Hansson <newbyte@postmarketos.org>
Date: Fri, 10 Oct 2025 12:12:10 +0200
Subject: [PATCH] colorchooserdialog: Make scrollable to fit narrow screens

Currently, the colour picker overflows on mobile phones. This is a
problem since there is no obvious replacement for libadwaita apps that
are meant to be adaptive, and while a proper replacement is planned for
sometime in the future, making the current dialogue at least fit narrow
screens works as a stopgap solution until then.

Related: https://gitlab.gnome.org/World/Phosh/phosh-mobile-settings/-/issues/87
Related: https://gitlab.gnome.org/GNOME/Initiatives/-/issues/13
Part-of: <https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/9029>
---
 gtk/gtkcolorchooserdialog.c     |  4 +++-
 gtk/ui/gtkcolorchooserdialog.ui | 32 ++++++++++++++++++++------------
 2 files changed, 23 insertions(+), 13 deletions(-)

diff --git a/gtk/gtkcolorchooserdialog.c b/gtk/gtkcolorchooserdialog.c
index 28e76f0bad3..3a403ab399e 100644
--- a/gtk/gtkcolorchooserdialog.c
+++ b/gtk/gtkcolorchooserdialog.c
@@ -66,6 +66,7 @@ struct _GtkColorChooserDialog
   GtkDialog parent_instance;
 
   GtkWidget *chooser;
+  GtkWidget *scroller;
 };
 
 struct _GtkColorChooserDialogClass
@@ -216,7 +217,7 @@ gtk_color_chooser_dialog_dispose (GObject *object)
 {
   GtkColorChooserDialog *cc = GTK_COLOR_CHOOSER_DIALOG (object);
 
-  g_clear_pointer (&cc->chooser, gtk_widget_unparent);
+  g_clear_pointer (&cc->scroller, gtk_widget_unparent);
 
   G_OBJECT_CLASS (gtk_color_chooser_dialog_parent_class)->dispose (object);
 }
@@ -252,6 +253,7 @@ gtk_color_chooser_dialog_class_init (GtkColorChooserDialogClass *class)
   gtk_widget_class_set_template_from_resource (widget_class,
 					       "/org/gtk/libgtk/ui/gtkcolorchooserdialog.ui");
   gtk_widget_class_bind_template_child (widget_class, GtkColorChooserDialog, chooser);
+  gtk_widget_class_bind_template_child (widget_class, GtkColorChooserDialog, scroller);
   gtk_widget_class_bind_template_callback (widget_class, propagate_notify);
   gtk_widget_class_bind_template_callback (widget_class, color_activated_cb);
 }
diff --git a/gtk/ui/gtkcolorchooserdialog.ui b/gtk/ui/gtkcolorchooserdialog.ui
index c8997d5d0e3..bb698459cca 100644
--- a/gtk/ui/gtkcolorchooserdialog.ui
+++ b/gtk/ui/gtkcolorchooserdialog.ui
@@ -2,7 +2,8 @@
 <interface domain="gtk40">
   <template class="GtkColorChooserDialog" parent="GtkDialog">
     <property name="title" translatable="yes">Select a Color</property>
-    <property name="resizable">0</property>
+    <property name="halign">GTK_ALIGN_CENTER</property>
+    <property name="valign">GTK_ALIGN_CENTER</property>
     <style>
       <class name="colorchooser"/>
     </style>
@@ -15,17 +16,24 @@
         <property name="margin-top">5</property>
         <property name="margin-bottom">5</property>
         <child>
-          <object class="GtkColorChooserWidget" id="chooser">
-            <property name="margin-start">5</property>
-            <property name="margin-end">5</property>
-            <property name="margin-top">5</property>
-            <property name="margin-bottom">5</property>
-            <property name="rgba">rgb(255,255,255)</property>
-            <property name="hexpand">1</property>
-            <property name="vexpand">1</property>
-            <signal name="color-activated" handler="color_activated_cb" swapped="no"/>
-            <signal name="notify::rgba" handler="propagate_notify" swapped="no"/>
-            <signal name="notify::show-editor" handler="propagate_notify" swapped="no"/>
+          <object class="GtkScrolledWindow" id="scroller">
+            <property name="max-content-height">262</property>
+            <property name="propagate-natural-width">1</property>
+            <property name="propagate-natural-height">1</property>
+            <child>
+              <object class="GtkColorChooserWidget" id="chooser">
+                <property name="margin-start">5</property>
+                <property name="margin-end">5</property>
+                <property name="margin-top">5</property>
+                <property name="margin-bottom">5</property>
+                <property name="rgba">rgb(255,255,255)</property>
+                <property name="hexpand">1</property>
+                <property name="vexpand">1</property>
+                <signal name="color-activated" handler="color_activated_cb" swapped="no"/>
+                <signal name="notify::rgba" handler="propagate_notify" swapped="no"/>
+                <signal name="notify::show-editor" handler="propagate_notify" swapped="no"/>
+              </object>
+            </child>
           </object>
         </child>
       </object>
-- 
GitLab

